Add hmac-sha256 support

dwmw2 opened the pull request:
# What does this implement/fix?

I was just adding hmac-md5 validation of OTA updates, and being sad about MD5... and realised we have SHA256 now. Here's HMAC support for it.

While at it, fix the existing sha256 tests which weren't running because we didn't define `USE_SHA256`, and add tests for HMAC-MD5 too.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Developer breaking change (an API change that could break external components)
- [ ] Code quality improvements to existing code or addition of tests
- [x] Other

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840
- [x] host
## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation not needed in [esphome-docs](https://github.com/esphome/esphome-docs).

dwmw2 committed:
Add HMAC-SHA256 component

Based on hmac_md5 component but using SHA256 instead of MD5.
Provides HMAC-SHA256 digest computation for cryptographic operations.

bdraco commented on esphome/components/hmac_sha256/hmac_sha256.cpp:
```diff
@@ -0,0 +1,58 @@
+#include "esphome/core/helpers.h"
+
+namespace esphome {
+namespace hmac_sha256 {
```
c++17 namespace please `esphome::hmac_sha256`

bdraco commented on esphome/components/hmac_sha256/hmac_sha256.cpp:
```diff
@@ -0,0 +1,58 @@
+namespace esphome {
+namespace hmac_sha256 {
+void HmacSHA256::init(const uint8_t *key, size_t len) {
+  uint8_t ipad[64], opad[64];
```
64 could be a constexpr instead of magic number

bdraco commented on esphome/components/hmac_sha256/hmac_sha256.h:
```diff
@@ -0,0 +1,49 @@
+  HmacSHA256() = default;
+  ~HmacSHA256() = default;
+
+  /// Initialize a new SHA256 digest computation.
```
```suggestion
  /// Initialize a new HMAC-SHA256 digest computation.
```

bdraco commented on esphome/components/hmac_sha256/hmac_sha256.h:
```diff
@@ -0,0 +1,49 @@
+  bool equals_hex(const char *expected);
+
+ protected:
+  sha256::SHA256 ihash_;
+  sha256::SHA256 ohash_;
```
Needs to be heavily tested on ESP32S3

see the warnings in esphome/components/sha256/sha256.cpp:13-44

dwmw2 replied:
Dropping ESP32 support for now; let's get the basic support in and we can resolve the ESP32-S3 weirdness separately.

bdraco replied:
I think it would be better to figure out esp32 first since the whole design may revolve around that.

bdraco replied:
We may need to use `mbedtls_md_hmac_` to avoid the dma issues on esp32s3

dwmw2 replied:
Added that for ESP32 and LibreTiny, both of which have mbedtls.

bdraco commented on esphome/components/sha256/__init__.py:
```diff
@@ -12,6 +12,8 @@
 
 
 async def to_code(config: ConfigType) -> None:
+    cg.add_define("USE_SHA256")
```
This looks like a bugfix ? Could be a separate PR

dwmw2 replied:
https://github.com/esphome/esphome/pull/12457

bdraco commented on esphome/components/md5/__init__.py:
```diff
@@ -1,7 +1,17 @@
 import esphome.codegen as cg
+from esphome.core import CORE
```
Please put the MD5 changes in a separate PR

dwmw2 replied:
- https://github.com/esphome/esphome/pull/12458
- https://github.com/esphome/esphome/pull/12459

dwmw2 committed:
Convert hmac_sha256 to C++17 nested namespace syntax

dwmw2 committed:
Use constexpr for HMAC block size with RFC 2104 reference

dwmw2 committed:
Update esphome/components/hmac_sha256/hmac_sha256.h

Co-authored-by: J. Nick Koston <nick+github@koston.org>

dwmw2 committed:
Fix signed/unsigned comparison warning in hmac_sha256

dwmw2 committed:
Drop ESP32 for hmac-sha256 for now

There is bizarre DMA-to-stack or similar weirdness on ESP32-S3, so disable
HMAC-SHA256 there for now until it can be tested/fixed. Let's get the
basic support infrastructure in first.

dwmw2 committed:
Merge branch 'dev' into hmac-sha256

dwmw2 committed:
Remove esp32 test

dwmw2 committed:
Add ESP32 support to hmac_sha256 using mbedtls HMAC

ESP32's SHA256 hardware acceleration has bugs when used across
function boundaries. Use mbedtls_md_hmac_* functions instead which
properly handle the hardware acceleration limitations.

dwmw2 committed:
Use SHA256_DIGEST_SIZE constant instead of hardcoded 32

dwmw2 committed:
Use SHA256_DIGEST_SIZE for hex string length in equals_hex()

dwmw2 committed:
Use mbedtls HMAC implementation for LibreTiny as well

dwmw2 committed:
Revert "Remove esp32 test"

This reverts commit bf831f213301ed5f8e344b331573bcbf0cd58ecb.

dwmw2 committed:
Add esp32-ard and rtl87xx-ard tests

dwmw2 committed:
Merge branch 'dev' into hmac-sha256

dwmw2 committed:
Merge branch 'dev' into hmac-sha256
